### PR TITLE
board/opentrons/ot2, configs/ot2_defconfig: Add rtc-rv3028 support

### LINKED ISSUE #124: feat: Add device tree overlay for rtc rv3028
- Write an overlay for the rv-3028 RTC. This will probably include just the rv-3028 fragment from the latest [upstream i2-rtc.dts](https://github.com/raspberrypi/linux/blob/9797f1a4938c20139b00a25de93cc99efb5c291b/arch/arm/boot/dts/overlays/i2c-rtc-overlay.dts) for this RTC. 
- Build it using dtc in buildroot’s post-image script. 
- Add the `dtoverlay` with name of this overlay file and the appropriate parameters.

To test the overlay in isolation, we can test it first with the new kernel that has the working driver for rv3028. Once verified that it works in the new kernel, we can use it with our current kernel and our back-ported device driver.

### LINKED ISSUE #125: feat: Backport the rtc-rv3028 driver
Modifying the driver source code ([this commit](https://github.com/raspberrypi/linux/blob/9797f1a4938c20139b00a25de93cc99efb5c291b/drivers/rtc/rtc-rv3028.c)) so that its dependencies are all in the 4.14.74 kernel.

Building the driver during the local kernel build:
- add the source file to the local repo
- add a Kconfig for the driver
- update the `drivers/rtc/makefile` to include the new driver
- include it in our board’s defconfig (`/arch/arm/configs/bcm2709_defconfig`) by adding the line `CONFIG_RTC_DRV_RV3028=m` (or `=y`)

Once we have a driver that compiles on 4.14.74, we will have to add patches for the above files in buildroot.

### PR DESCRIPTION
**Pull request by sanni-t**
Closes #124, Closes #125 

# overview

This PR adds rv3028 RTC support in our RobotOS; necessary for shipping refresh robots that have the RTC added.

NOTE: Requires #126 to be fully functional. The PR for it will be merged into this one once the two are tested separately.

# changes
- Adds a device tree overlay for rv3028 + enables the overlay using `dtoverlay` in config.txt
- Adds kernel patches for adding the rv3028 driver. 
    - Commits are cherry picked from [this file history](https://github.com/raspberrypi/linux/commits/feb909daf012ea48171d6b65ea46614ec524c81e/drivers/rtc/rtc-rv3028.c). We skip over commit #(44c638c) since it does only a trivial refactor of rv3028 driver while changing other files that are irrelevant for us.
    - Last two commits:
        1. backports the rtc driver to run on our current kernel (4.14)- thanks to @amitlissack! 
        2. configures to build the driver in kernel
- Enables linux utilities- i2ctools (for i2c debugging) & hwclock (to handle rtc read/write)

# review request:

Build links:
- [ot2-fullImage.zip](https://opentrons-buildroot-ci.s3.amazonaws.com/2e1c3f98-a2b4-46df-9ba0-79674f455353/opentrons-buildroot/ot2-fullimage.zip)
- [ot2-system.zip](https://opentrons-buildroot-ci.s3.amazonaws.com/2e1c3f98-a2b4-46df-9ba0-79674f455353/opentrons-buildroot/ot2-system.zip)

1. On a refresh robot, this build should enable RTC use. Test that:
    - The rtc is detected and registered by the kernel: `dmesg | grep rtc` should show `rtc-rv3028 1-0052: registered as rtc0`.
    - Backup switchover mode is set correctly: 
    ```
    modprobe i2c-dev
    i2cdump -f -y -r 0x37-0x37 1 0x52 b
    ```
    Should output a register map showing value of reg `0x37` as `94` or `14`
    - If this is the first boot, and your robot doesn't have an internet connection, and your RTC wasn't previously set, update the RTC with *current system time* using `hwclock -w --noadjfile --utc`. You can check that it was set correctly with `hwclock -r`
    - Connect to the internet if you haven't already and wait for systemd-timesyncd to sync with NTP. You can check the sync status with `timedatectl`. Once synced, `timedatectl` should show your RTC as updated to the correct UTC time and should be same as system clock (difference of a few milliseconds is fine).
    - Disconnect from the internet and power off the robot. After about 10-15 minutes, switch the robot back on without internet. `timedatectl` should still show the correct RTC time but system clock will be behind by 10-15 mins.

2. On a non refresh robot, use the ot2-system.zip file to send this update through the update server. Once updated, grep for `rtc` in system logs and see that there are no errors found. Also make sure that our current time sync infrastructure (time sync on app connection & timer scripts) still work correctly.

### CODE REVIEW

File: board/opentrons/ot2/kernel-patches/0007-Backport-rv3028-driver-for-4.14-kernel.patch [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+-	rv3028->rtc->range_max = RTC_TIMESTAMP_END_2099;
++	//rv3028->rtc->range_min = RTC_TIMESTAMP_BEGIN_2000;
++	//rv3028->rtc->range_max = RTC_TIMESTAMP_END_2099;
+ 	rv3028->rtc->ops = &rv3028_rtc_ops;
+ 	ret = rtc_register_device(rv3028->rtc);
+ 	if (ret)
+ 		return ret;
+-
++/*
+ 	nvmem_cfg.priv = rv3028->regmap;
```
Comment (amitlissack) [inline board/opentrons/ot2/kernel-patches/0007-Backport-rv3028-driver-for-4.14-kernel.patch:82]: Really happy we didn't end up needing the code I wrote as a replacement for these `rtc_nvmem_register` function calls.

**Review by amitlissack — APPROVED** [review]

File: board/opentrons/ot2/post-image.sh [hunk 2]
```diff
@@ -60,6 +60,11 @@ cat << __EOF__ >> "${BINARIES_DIR}/rpi-firmware/config.txt"
 dtoverlay=gpio-revision-bits
 __EOF__
 
+dtc -@ -I dts -O dtb -o "${BINARIES_DIR}/rpi-firmware/overlays/i2c-rtc-rv3028.dtbo" "${BOARD_DIR}/i2c-rtc-rv3028.dts"
+cat << __EOF__ >> "${BINARIES_DIR}/rpi-firmware/config.txt"
+dtoverlay=i2c-rtc-rv3028
+__EOF__
+
```
Comment (SyntaxColoring) [inline board/opentrons/ot2/post-image.sh:67]: I know you're just continuing the pattern from the code directly above this, but what's the difference between this code and whatever is going on in genimage-ot2.cfg?

https://github.com/Opentrons/buildroot/blob/4f575ca1e3c0449666229871be8897206cda14d2/board/opentrons/ot2/genimage-ot2.cfg#L1-L18

File: board/opentrons/ot2/kernel-patches/0008-build-rv3028-driver-into-kernel.patch [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
++++ b/arch/arm/configs/bcm2709_defconfig
+@@ -1146,6 +1146,7 @@ CONFIG_RTC_DRV_PCF2123=m
+ CONFIG_RTC_DRV_DS3232=m
+ CONFIG_RTC_DRV_PCF2127=m
+ CONFIG_RTC_DRV_RV3029C2=m
++CONFIG_RTC_DRV_RV3028=y
+ CONFIG_DMADEVICES=y
+ CONFIG_DMA_BCM2835=y
+ CONFIG_DMA_BCM2708=y
+--
```
Comment (SyntaxColoring) [inline board/opentrons/ot2/kernel-patches/0008-build-rv3028-driver-into-kernel.patch]: Modifying kernel configuration through a patch file feels indirect to me. Like it competes with buildroot's own build system.

By comparison, here's what Orange Pi does:

https://github.com/Opentrons/buildroot/blob/4f575ca1e3c0449666229871be8897206cda14d2/configs/orangepi_lite_defconfig#L11

Would that make more sense? I have a hard time following Buildroot's build system, so maybe I'm missing something.

**Review by SyntaxColoring — COMMENTED** [review]
Just looking at the code so far. Haven't tested on my non-refresh OT-2, but I will.

File: board/opentrons/ot2/post-image.sh (hunk 2 above)
Comment (sanni-t) [reply board/opentrons/ot2/post-image.sh:67]: The `post-image.sh` code compiles the device tree overlay source file `i2c-rtc-rv3028.dts` to produce a device tree overlay binary `i2c-rtc-rv3028.dtbo` in the overlays directory mentioned in `genimage-ot2.cfg`. It also adds the `dtoverlay=i2c-rtc-rv3028` line to config.txt. That's what enables the overlay. 
The genimage file provides info on the partition layout to the genimage tool that creates our .img file.

Comment (sfoster1) [reply board/opentrons/ot2/post-image.sh:67]: the genimage script copies in previously-existing files unchanged. the post-image script is a) compiling the dt overlay b) actualy editing the config.txt file.

there's probably a better way to do the stuff in the post-image script, like by adding drop ins or modifications to the rpi firmware package, but this works for now.

File: board/opentrons/ot2/kernel-patches/0008-build-rv3028-driver-into-kernel.patch (hunk 3 above)
Comment (sfoster1) [reply board/opentrons/ot2/kernel-patches/0008-build-rv3028-driver-into-kernel.patch]: Yes, that would work and is probably a better way to do it. Really the outright correct thing to do is make our own kernel fork and a dedicated kernel config, but the fragments approach is fine too.

**Review by sfoster1 — APPROVED** [review]
Code and config changes look good! Don't have a refresh but it boots fine on my non-refresh.

**Review by SyntaxColoring — APPROVED** [review]
> On a non refresh robot, use the ot2-system.zip file to send this update through the update server. Once updated, grep for rtc in system logs and see that there are no errors found. Also make sure that our current time sync infrastructure (time sync on app connection & timer scripts) still work correctly.

Confirmed all of this. 🎉

File: board/opentrons/ot2/kernel-patches/0008-build-rv3028-driver-into-kernel.patch (hunk 3 above)
Comment (sanni-t) [reply board/opentrons/ot2/kernel-patches/0008-build-rv3028-driver-into-kernel.patch]: Testing this change

**Review by mcous — APPROVED** [review]
Tested on OT-2 Refresh hardware via the #129 branch

🕐 🕑 🕒 🕓 🕔 🕕 🕖 🕗 🕘 🕙 🕚 🕛

**Pull request merged**